odoo/addons/base/models/res_lang.py

```diff
@@ -6,17 +6,40 @@
 import locale
 import logging
 import re
-from operator import itemgetter
+from typing import Any, Literal
 
 from odoo import api, fields, models, tools, _
 from odoo.exceptions import UserError, ValidationError
+from odoo.tools import OrderedSet
 
 _logger = logging.getLogger(__name__)
 
 DEFAULT_DATE_FORMAT = '%m/%d/%Y'
 DEFAULT_TIME_FORMAT = '%H:%M:%S'
 
 
+class LangData(tools.frozendict):
+    """ A data class which can access field value like a res.lang record
+    e.g. lang_data.code
+    """
+    __slots__ = ()
+
+    def __getattribute__(self, field: str) -> Any:
+        return self[field]
```

Comment:
Note: Why we override `__getattribute__` instead of `__getattr__`

If we override `__getattr__`, `LangData` will have all public methods inherited from `dict`.
1. if a field name is the same as a `dict`'s method name e.g `keys` 😂 you cannot cache the field. because `lang_data.keys` will return a function
(while field names like dict`'s method names are always not good idea in practice and in res.lang)
2. if a developer forgets it is a `LangData` object and use it like a record, and call `lang_data.copy()`. The real record won't be copied. And there is no error raised
(while `res_lang.copy()` is always wrong for sake of the unique constaints, and you shouldn't write code like this)

We shouldn't have all those corner cases in Odoo
But as a low level class, it is not bad to make its feature less tricky.